libtiledbsoma/src/soma/soma_array.cc

```diff
@@ -1178,81 +1178,116 @@ uint64_t SOMAArray::nnz_slow() {
-                break;
-            case TILEDB_UINT16:
-                result.push_back(
-                    dim.domain<uint16_t>().second -
-                    dim.domain<uint16_t>().first + 1);
-                break;
-            case TILEDB_INT16:
-                result.push_back(
-                    dim.domain<int16_t>().second - dim.domain<int16_t>().first +
-                    1);
-                break;
-            case TILEDB_UINT32:
-                result.push_back(
-                    dim.domain<uint32_t>().second -
-                    dim.domain<uint32_t>().first + 1);
-                break;
-            case TILEDB_INT32:
-                result.push_back(
-                    dim.domain<int32_t>().second - dim.domain<int32_t>().first +
-                    1);
-                break;
-            case TILEDB_UINT64:
-                result.push_back(
-                    dim.domain<uint64_t>().second -
-                    dim.domain<uint64_t>().first + 1);
-                break;
-            case TILEDB_INT64:
-            case TILEDB_DATETIME_YEAR:
-            case TILEDB_DATETIME_MONTH:
-            case TILEDB_DATETIME_WEEK:
-            case TILEDB_DATETIME_DAY:
-            case TILEDB_DATETIME_HR:
-            case TILEDB_DATETIME_MIN:
-            case TILEDB_DATETIME_SEC:
-            case TILEDB_DATETIME_MS:
-            case TILEDB_DATETIME_US:
-            case TILEDB_DATETIME_NS:
-            case TILEDB_DATETIME_PS:
-            case TILEDB_DATETIME_FS:
-            case TILEDB_DATETIME_AS:
-            case TILEDB_TIME_HR:
-            case TILEDB_TIME_MIN:
-            case TILEDB_TIME_SEC:
-            case TILEDB_TIME_MS:
-            case TILEDB_TIME_US:
-            case TILEDB_TIME_NS:
-            case TILEDB_TIME_PS:
-            case TILEDB_TIME_FS:
-            case TILEDB_TIME_AS:
-                result.push_back(
-                    dim.domain<int64_t>().second - dim.domain<int64_t>().first +
-                    1);
-                break;
-            default:
-                throw TileDBSOMAError("Dimension must be integer type.");
+        // Callers inquiring about non-int64 shapes should not be here.
+        //
+        // In the SOMA data model:
+        // * SparseNDArray has dims which are all necessarily int64_t
+        // * DenseNDArray has dims which are all necessarily int64_t
+        // * DataFrame _default_ indexing is one dim named "soma_dim_0" of type
+        //   int64_t, however:
+        //   * Users can (and do) add other additional dims
+        //   * The SOMA data model requires that soma_joinid be present in each
+        //     DataFrame either as a dim or an attr -- and there are DataFrame
+        //     objects for which soma_joinid is not a dim at all
+        //   * These cases are all actively unit-tested within apis/python/tests
+        if (dim.type() != TILEDB_INT64) {
+            throw TileDBSOMAError("Found unexpected non-int64 dimension type.");
```

Comment:
I know where you are coming from here, and I slept over it one night, but I am not sure I am fully on-board. 

We have the factory methods.  We have the SOMA definitions.  It is clear what a well-formed SOMA object is.  

But I  still cannot shake the feeling that this seems "wrong" in the sense of working, testing, testable, useful, ... general code.  

I'll sleep over it another night.  Not objecting to the change, but scratching my chin.  :thinking:

Reply:
The removed code is dead ... unit tests would be failing if it were not. :)

Reply:
That was not really my point / I understand that.  But such a converter is also useful outside of SOMA, the test is fine, and I have tested and developed too with non-soma dataframes.  That can be useful.  

Again, not vetoing anything here, just saying that I simultaneously understand why you propose what you and that I likely would not do it.  No more, no less, and surely no veto.

Reply:
Also "we can remove because doing so does not break tests" is not really a viable rule because we all know that coverage is never perfect.  Anyway ...